### PR TITLE
Create attack state

### PR DESCRIPTION
**Pull request by dedeswim**
This PR introduces a state for the attack so that the results can be restored in case the evaluation gets interrupted, and the state can be resumed (e.g., because the resources used for the evaluation are preempted).

The state is a `dataclass` which saves the following information:

- the path where the state is saved on disk
- the attacks that have been run so far
- a `bool` for each image to attack, where `False` means that the image has already been successfully attacked
- the time when the state was last saved to disk. This is to avoid bloating I/O by saving the state too often. By using this, the state is saved at most once every 60s.
- the clean accuracy computed at the beginning of the evaluation

It's not necessary to save the robust accuracy, as it can be easily computed by doing `torch.sum(robust_flags).item() / robust_flags.shape[0]`.

To save and restore the state, it's enough to pass the path where the state has to be (or is already) saved to `AutoAttack.run_standard_evaluation`.

### CODE REVIEW

**Comment by dedeswim** [comment]
Hi! Thanks for the reply. Just to check: you mean to add a warning in the example script, i.e., [here](https://github.com/dedeswim/auto-attack/blob/master/autoattack/examples/eval.py#L71)?

**Comment by dedeswim** [comment]
Meanwhile, I added it, let me know if this is what you meant 😊

**Comment by fra31** [comment]
I was rather thinking to somewhere like [here](https://github.com/dedeswim/auto-attack/blob/fc218ea43409c8f66af9c66379f80943ad4fcfe0/autoattack/autoattack.py#L95) (at the beginning if `state_path` is not `None`) or [here](https://github.com/dedeswim/auto-attack/blob/fc218ea43409c8f66af9c66379f80943ad4fcfe0/autoattack/autoattack.py#L241) (at the end of the evaluation if `state_path` is not `None`), maybe both. I would add it in `autoattack.py` itself, so that it's independent of the example. Does it make sense to you?

The message looks good to me, just a minor edit below, I'd just use `self.logger.log` to print them, so that they get potentially saved to the main log file as well (similarly for [this](https://github.com/dedeswim/auto-attack/blob/adfb4e5e068df089a1a068560fe7d687e6471396/autoattack/autoattack.py#L90)).
```
Since a state path is provided, the saved adversarial examples are **only** those obtained in the latest run of the attack.
```

**Comment by dedeswim** [comment]
I see. Then I would refer not to the saved adversarial examples, but to those returned by the attack. What do you think?

**Comment by dedeswim** [comment]
Fixed. I have also changed a couple more `print` statements I found to `self.logger.log`. Would it make sense to mention in the warning that adversarial accuracy can be computed with `state.robust_flags.mean()`? Or should I create a `@property` method that computes it?

**Comment by dedeswim** [comment]
I added the property, but if a user calls it before all the attacks have been run, then it warns the user that they are checking it before all the attacks are finished. Moreover, I also added a check that the attacks to run match between the AutoAttack object and the restored state. This should prevent some errors by users loading the wrong state file

**Comment by fra31** [comment]
Great, thanks again!
